projects/ADuCM3029_demo_cn0415/src/cn0415.c

```diff
@@ -0,0 +1,2189 @@
+/******************************************************************************/
+extern float dith_ampl;
+extern uint8_t uart_current_line[256];
+extern uint8_t uart_previous_line[256];
+extern uint8_t uart_line_index;
+extern uint8_t uart_cmd;
+extern uint8_t pid_flag;
+extern bool dither_period;
+
+/* Available commands */
+char *cmd_commands[] = {
+	"help",
+	"h",
+	"frequency ",
+	"f ",
+	"duty_cycle ",
+	"d ",
+	"read_adc",
+	"r",
+	"activate_dither",
+	"ad",
+	"remove_dither",
+	"rd",
+	"dither_frequency ",
+	"df ",
+	"dither_amplitude ",
+	"da ",
+	"set_pid_freq ",
+	"spf ",
+	"set_pid_sp ",
+	"sps ",
+	"add_pid",
+	"ap",
+	"remove_pid",
+	"rp",
+	"set_pid_kp ",
+	"kp ",
+	"set_pid_ki ",
+	"ki ",
+	"set_pid_kd ",
+	"kd ",
+	"set_pulse_value ",
+	"spuv ",
+	"set_pulse_period ",
+	"spup ",
+	"dig_sol_on",
+	"dson",
+	"dig_sol_off",
+	"dsof",
+	"prop_sol_set ",
+	"pss ",
+	"reset",
+	"rst",
+	"status",
+	"stts",
+	"calibrate",
+	"clb",
+	""
+};
+
+/* Functions for available commands */
+cmd_func v_cmd_fun[] = {
+	cn0415_help,
+	cn0415_pwm_set_freq,
+	cn0415_pwm_set_duty,
+	cn0415_read_voltage,
+	cn0415_dither_activate,
+	cn0415_dither_remove,
+	cn0415_dither_set_freq,
+	cn0415_dither_set_ampl,
+	cn0415_pid_set_sample_frequency,
+	cn0415_pid_set_setpoint,
+	cn0415_pid_add,
+	cn0415_pid_remove,
+	cn0415_pid_set_kp,
+	cn0415_pid_set_ki,
+	cn0415_pid_set_kd,
+	cn0415_pulse_set_val,
+	cn0415_pulse_set_period,
+	cn0415_dig_sol_on,
+	cn0415_dig_sol_off,
+	cn0415_prop_sol_set,
+	cn0415_reset,
+	cn0415_status,
+	cn0415_calibration,
+	NULL
+};
+
+uint8_t command_size[] = {5, 2, 10, 2, 11, 2, 9, 2, 16, 3, 14, 3, 17, 3, 17, 3,
+			  13, 4, 11, 4, 8, 3, 11, 3, 11, 3, 11, 3, 11, 3, 16, 5, 17, 5, 11,
+			  5, 12, 5, 13, 4, 6, 4, 7, 5, 10, 4, 1
+			 };
+
+float offset_err, gain_err = 0;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/***************************************************************************//**
+ * @brief Initializes the cn0415 device.
+ *
+ * @param device     - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       		   parameters.
+ *
+ * @return ret - Result of the initialization procedure.
+ *               Example: 0 - if initialization was successful;
+ *                        -1 - if initialization was unsuccessful.
+*******************************************************************************/
+int32_t cn0415_setup(struct cn0415_dev **device,
+		     struct cn0415_init_param init_param)
+{
+	struct cn0415_dev *dev;
+	int32_t ret;
+
+	/* Allocate memory for board descriptor */
+	dev = (struct cn0415_dev *)malloc(sizeof(*dev));
+	if(!dev)
+		return -1;
+
+	/* Initialize Systick timer */
+	ret = timer_start();
+	if(ret != 0)
+		goto error;
+#ifdef FLASH_INIT_PARAMETERS
```

Comment:
Curious: do we really have to use conditional groups?

Reply:
I only did it this way to follow the example projects in the no-OS repo. If there is a better way to do it I want to do that.

Reply:
We need the program to have ways to initialize the application: from the main function in the ADuCM_demo_cn0415.c or from the flash. Also to have a easy way to change the flash initialization values.